Effects

Sorien opened the pull request:
![image](https://user-images.githubusercontent.com/1270275/40740313-cb7cbfa2-6448-11e8-9fea-fb09b4935856.png)

- What does the pull request do?

implements DropShadow effect (Skia, D2D), only for deferred rendering, i wanted to have it more like HTML than WPF so there are different properties

- How was the solution implemented (if it's not obvious)?

visual that contains effect will be rendered to separate layer, deferred renderer has one more layer (_effectLayer) that is used for applying effect (d2d cannot draw with a bitmap that is currently bound as the target bitmap) - don't like it and don't know how to solve it, this layer is cleared before applying clipping and effect otherwise there are visual artifacts

there are image page changes that should be reverted, also there are no unit tests, _i won't have a time to finish_ it so if anybody want to finish it you have my blessing

Checklist:

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

Fixes #1501

nc4rrillo commented:
Are you planning on adding offset, blur, color and spread properties to the drop shadow effect?

Commit:
Effects

Sorien commented:
i'll try to implement shadow and blur effect (skia and D2D) with all properties, but at first have to find right design, it's a bit hacky

nc4rrillo commented:
Would it be possible to have this extend DrawingContext with something like `PushDropShadow(...)`/`PopDropShadow()`? that way you don't need the `IEffectImpl` and can just leverage the existing `DrawingContext` infrastructure

Gillibald commented:
In my opinion an effect is just some manipulation to the rendered image. This should be implemented like a pipeline and transparent to all render calls. That way you can have several manipulations (Filter, Transform, etc) on a generated image. As far as I know Direct2D has a concept of composable effects. So you can have more than one effect like shadow blur etc. Each is an effect on its own.

Gillibald commented:
I just thought a bit about this PR and checked how skia handles effects (filters). Filters are always applied to SkPaint in our case all drawings written to SkPaint (Canvas).

Can we add effects to our drawing context interface and use them for every drawing call?

To use multiple effects we can use a combined effect or a list of effects. I much prefer having a list and combine them if needed in the implementation.

Commit:
do not apply effect during layer compositing 

redraw whole layer with Effect
apply proper clipping

Commit:
remove IEffectImpl  from DrawImage

Commit:
shadow effect properties

Commit:
invalidate visual when effect's properties changes

Commit:
clean up

Commit:
D2D shadow implementation 

d2d needs special layer - cannot draw with a bitmap that is currently bound as the target bitmap

nc4rrillo commented:
Perhaps `PushOpacity` should become `PushFilters` with Opacity becoming one of many filters than can be pushed, others being DropShadow, Blur, Sepia, Grayscale, Invert, etc.

Commit:
add missing SharpDX package

Commit:
simplify effect drawing

jkoritzinsky commented on build/SharpDX.props:
```diff
@@ -4,6 +4,7 @@
     <PackageReference Include="SharpDX.Direct2D1" Version="4.0.1" />
     <PackageReference Include="SharpDX.Direct3D11" Version="4.0.1" />
     <PackageReference Include="SharpDX.DXGI" Version="4.0.1" />
+    <PackageReference Include="SharpDX.Mathematics" Version="4.0.1" />
```
Make sure to add this package reference to the packages.cake build script.

ryuujo1573 commented:
Sorry I'm not good at English.. It might be rude but excuse me, I was wondering if this feature is implemented and if so how could I add it to a simple Border with adorable shadow effect?

Gillibald commented:
BoxShadows are supported in the current master

ryuujo1573 commented:
Thank you so much!! It's so surprising to be replied so soon like this! THX!!

graealex commented:
I'm a bid saddened that this effects pipeline never made it into a release. BoxShadow is nice, if you want a shadow. But not if you want to blur something.
